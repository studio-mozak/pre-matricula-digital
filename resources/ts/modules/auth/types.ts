export interface Auth {
  level: number;
  name: string;
  schools: string[];
}
