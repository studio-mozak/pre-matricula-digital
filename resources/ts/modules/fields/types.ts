import { Nullable } from '@/types';

export type SetValues = (fields: Record<string, unknown>) => void;

export interface FieldPageOption {
  id?: string;
  name: Nullable<string>;
  weight: number;
}

export interface FieldPage {
  id: Nullable<string>;
  name: Nullable<string>;
  description: Nullable<string>;
  group: Nullable<'RESPONSIBLE' | 'STUDENT'>;
  type: Nullable<string>;
  internal: boolean;
  options: FieldPageOption[];
  deleteOptions?: string[];
  required?: boolean;
}

export interface ErrorResponse {
  message: string;
}
