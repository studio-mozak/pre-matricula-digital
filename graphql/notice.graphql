extend type Query @guard(with: ["web", "sanctum", "prematricula"]) {

  """
  Lista todos os avisos.
  """
  notices: [Notice!]! @paginate

  notice(id: ID @eq): Notice @find

}

extend type Mutation @guard {

  saveNotice(
    input: NoticeInput! @spread
  ): Notice!

  deleteNotice(
    id: ID! @whereKey
  ): Notice! @delete
}

type Notice {
  id: ID!
  text: String!
}

input NoticeInput {
  id: ID
  text: String!
}
