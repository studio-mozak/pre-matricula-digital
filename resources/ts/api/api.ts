import axios, { AxiosInstance } from 'axios';

const token = window.config ? window.config.token : '';

export const http: AxiosInstance = axios.create({
  baseURL: String(import.meta.env.VITE_APP_API_URL || ''),
  headers: {
    Authorization: `Token ${token}`,
  },
});

axios.interceptors.response.use(
  (res) => {
    /**
     * If the response from the server has the "errors" object or the object "data" is empty,
     * here it is forced to go to the catch block automatically.
     */
    if (
      (res.data.errors || Object.keys(res.data).length === 0) &&
      !(res.data instanceof Blob)
    ) {
      return Promise.reject({
        response: {
          data: res.data,
        },
      });
    }

    return res;
  },
  (err) => Promise.reject(err)
);

export const rest = <T>(
  type: 'get' | 'post' | 'put' | 'delete',
  url: string,
  params?: object
): Promise<T> => http[type](url, params);

export const graphql = <T>(data: object, config?: object): Promise<T> =>
  axios.post('/graphql', data, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
    ...config,
  });
