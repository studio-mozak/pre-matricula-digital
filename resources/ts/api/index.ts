export * from './api';
export * as City from './services/graphql/city';
export * as Resource from './services/graphql/resource';
