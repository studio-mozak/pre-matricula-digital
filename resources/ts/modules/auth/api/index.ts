/**
 * Exporting the Axios Instance
 */
export { http, graphql } from '@/api';

/**
 * Exporting all modules
 */
export * as AuthRest from '@/modules/auth/api/services/rest/auth';
