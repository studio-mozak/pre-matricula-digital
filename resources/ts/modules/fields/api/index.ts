/**
 * Exporting the Axios Instance
 */
export { http, graphql } from '@/api';

/**
 * Exporting all modules
 */
export * as Field from './services/graphql/field';
