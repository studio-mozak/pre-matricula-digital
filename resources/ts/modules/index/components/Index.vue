<template>
  <div>
    <div class="row mt-3">
      <header-entity-data />
    </div>
    <div class="row mt-3">
      <div class="col-12 col-md-10 offset-md-1">
        <protocol-consult title="Consulta de inscrição" />
      </div>
    </div>
    <skeleton-card-notices v-if="loadingNotices" />
    <div
      v-if="hasNotice && !loadingNotices"
      data-test="notice-block"
      class="row mt-4"
    >
      <div class="col-12 col-md-10 offset-md-1">
        <x-card bordered>
          <x-card-section class="pl-4 pr-4 pt-3 p-0">
            <h5 class="text-h5 text-primary">Avisos</h5>
          </x-card-section>
          <x-card-section class="pl-4 pr-4 pb-3 p-0">
            <div class="row">
              <div
                class="col-12 text-break overfflow-hidden"
                v-html="truncateText"
              ></div>
            </div>
            <div class="row">
              <div
                class="col-11 col-lg-2 font-hind offset-lg-10 pr-0 pr-md-4 text-right"
              >
                <router-link to="/avisos">Leia mais</router-link>
              </div>
            </div>
          </x-card-section>
        </x-card>
      </div>
    </div>
    <skeleton-card-processes v-if="loadingProcesses" />
    <div v-else class="row mt-3">
      <div class="container-sm col-12 col-md-10 offset-md-1 mt-4">
        <div class="row">
          <div class="col">
            <h2 class="title mb-3">Inscrições</h2>
          </div>
        </div>
        <div
          v-for="process in processes"
          :key="process.id"
          data-test="process-block"
        >
          <div class="row">
            <div class="col-12 mt-3 mt-sm-2 d-flex">
              <h4 class="m-0">
                {{ process.name }} ({{ process.schoolYear.year }})
              </h4>
              <router-link
                v-if="process.showWaitingList"
                :to="`/lista-de-espera/${process.id}`"
                class="ml-auto p-2 bg-primary rounded text-white"
              >
                Consultar lista de espera
              </router-link>
              <span
                v-else
                class="ml-auto p-2 bg-gray-300 text-gray-700 rounded"
              >
                Lista de espera não disponível
              </span>
            </div>
          </div>
          <process-stages :process="process" class="mb-3" />
        </div>
      </div>
    </div>
    <modal
      v-model="showVideoTutorial"
      data-test="modal-intro"
      title="Boas-vindas!"
      :large="true"
    >
      <template #body>
        <div class="row">
          <div class="col-12 col-lg-8 offset-lg-2 text-center">
            <span>
              Olá! Você sabe como realizar uma solicitação de matrícula? Caso
              ainda não saiba, assista o vídeo abaixo e veja como realizar a
              inscrição.
            </span>
          </div>
        </div>
        <div class="row mt-4">
          <div
            class="col-12 col-lg-8 offset-lg-2 embed-responsive embed-responsive-16by9"
          >
            <iframe
              class="embed-responsive-item"
              :src="getConfig.video_intro_url"
              allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
              allowfullscreen
              title="Vídeo instrutivo"
            >
            </iframe>
          </div>
        </div>
      </template>
      <template #footer>
        <div class="d-flex flex-md-row flex-column justify-content-center">
          <div
            class="d-flex align-items-center justify-content-center justify-content-lg-start"
          >
            <div class="custom-control custom-checkbox">
              <input
                id="showVideoTutorial"
                v-model="dontShowVideoTutorial"
                type="checkbox"
                class="custom-control-input"
              />
              <label
                for="showVideoTutorial"
                class="custom-control-label font-weight-bold"
              >
                Não mostrar mais esta mensagem
              </label>
            </div>
          </div>
          <div class="d-flex align-items-center ml-0 ml-lg-3 mt-3 mt-lg-0">
            <x-btn
              label="Entendi"
              class="w-100"
              color="primary"
              no-caps
              no-wrap
              @click="close"
            />
          </div>
        </div>
      </template>
    </modal>
  </div>
</template>

<script lang="ts">
import { Processes, Stages, Vacancy } from '@/types';
import { Ref, computed, defineComponent, reactive, ref } from 'vue';
import { getCookie, setCookie } from '@/util';
import HeaderEntityData from '@/components/elements/HeaderEntityData.vue';
import Modal from '@/components/elements/Modal.vue';
import { Notice } from '@/modules/notice/types';
import { Notice as NoticeApi } from '@/modules/notice/api';
import { Process as ProcessApi } from '@/modules/processes/api';
import ProcessStages from '@/components/elements/ProcessStages.vue';
import ProtocolConsult from '@/components/elements/ProtocolConsult.vue';
import SkeletonCardNotices from '@/components/loaders/components/CardNotices.vue';
import SkeletonCardProcesses from '@/components/loaders/components/CardProcesses.vue';
import XBtn from '@/components/elements/buttons/XBtn.vue';
import XCard from '@/components/elements/cards/XCard.vue';
import XCardSection from '@/components/elements/cards/XCardSection.vue';
import clip from 'text-clipper';
import { useGeneralStore } from '@/store/general';
import { useLoader } from '@/composables';
export default defineComponent({
  components: {
    Modal,
    HeaderEntityData,
    ProtocolConsult,
    ProcessStages,
    SkeletonCardNotices,
    SkeletonCardProcesses,
    XBtn,
    XCard,
    XCardSection,
  },
  setup() {
    const vacancies: Vacancy[] = reactive([]);
    const showVideoTutorial: Ref<boolean> = ref(false);
    const dontShowVideoTutorial: Ref<boolean> = ref(false);
    const cookieName: Ref<string> = ref('SHOW_VIDEO_INTRO');

    const store = useGeneralStore();
    const {
      data: notice,
      loading: loadingNotices,
      loader: loadNotice,
    } = useLoader<Notice>();
    const {
      loading: loadingProcesses,
      loader: loadProcess,
      data: processes,
    } = useLoader<Processes[]>([]);

    const getProcesses = computed<Processes[]>(() => {
      const processesOpened: Processes[] = [];
      const processesNotOpened: Processes[] = [];
      processes.value.forEach((process: Processes) => {
        if (process.stages.some((stage: Stages) => stage.status === 'OPEN')) {
          processesOpened.push(process);
        } else {
          processesNotOpened.push(process);
        }
      });
      const processesOpenedWithDate = processesOpened.map((process) => {
        const { stages } = process;
        const internalProcess: Processes = { ...process };
        stages.sort((a: Stages, b: Stages) => {
          const aDate = new Date(a.startAt);
          const bDate = new Date(b.startAt);
          /**
           * Segundo este link, uma boa solução para subtrair valores diferentes de números, seria colocando o operador
           * "+" na frente das variáveis que desejamos calcular.
           *
           * https://stackoverflow.com/questions/14980014/how-can-i-calculate-the-time-between-2-dates-in-typescript
           */
          return +bDate - +aDate;
        });
        internalProcess.startAt = stages[0].startAt;
        return internalProcess;
      });
      const processesNotOpenedWithDate = processesNotOpened.map((process) => {
        const { stages } = process;
        const internalProcess = { ...process };
        stages.sort((a, b) => +new Date(b.startAt) - +new Date(a.startAt));
        internalProcess.startAt = stages[0].startAt;
        return internalProcess;
      });
      processesOpenedWithDate.sortBy('startAt').reverse();
      processesNotOpenedWithDate.sortBy('startAt').reverse();
      return processesOpenedWithDate.concat(processesNotOpenedWithDate);
    });
    const truncateText = computed(() =>
      notice.value.text
        ? clip(notice.value.text, 193, { html: true, maxLines: 4 })
        : ''
    );
    const hasNotice = computed(() => !!notice.value && !!notice.value.id);
    const getConfig = computed(() => store.getConfig);
    const showHowToDoVideo = computed(
      () => getConfig.value.show_how_to_do_video
    );
    function close() {
      if (dontShowVideoTutorial.value) {
        setCookie(cookieName.value, '1', 365);
      }
      showVideoTutorial.value = false;
    }
    async function getNotices() {
      loadNotice(NoticeApi.list);
    }
    async function getProceses() {
      loadProcess(ProcessApi.list).then((res) => {
        const allProcesses = res.filter((process) => process.stages.length);

        processes.value = allProcesses;
      });
    }
    function getData() {
      getNotices();
      getProceses();
    }
    getData();
    if (getCookie(cookieName.value) !== '1' && showHowToDoVideo.value) {
      showVideoTutorial.value = true;
    }

    return {
      vacancies,
      loadingNotices,
      loadingProcesses,
      showVideoTutorial,
      dontShowVideoTutorial,
      cookieName,
      notice,
      processes: getProcesses,
      truncateText,
      hasNotice,
      close,
      getConfig,
      showHowToDoVideo,
    };
  },
});
</script>
