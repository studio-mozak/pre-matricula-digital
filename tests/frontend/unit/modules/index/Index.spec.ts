import '@/plugin/sortby';
import { VueWrapper, flushPromises, mount } from '@vue/test-utils';
import Index from '@/modules/index/components/Index.vue';
import { Notice } from '@/modules/notice/types';
import { Notice as NoticeApi } from '@/modules/notice/api';
import { Process as ProcessApi } from '@/modules/processes/api';
import { Processes } from '@/types';
import VTooltip from 'v-tooltip';
import { createTestingPinia } from '@pinia/testing';
import { getCookie } from '@/util';
import { expect, vi } from 'vitest';

const spyPush = vi.fn();

vi.mock('vue-router/dist/vue-router.mjs', () => ({
  useRouter: () => ({
    push: spyPush,
  }),
}));
vi.mock('@/modules/notice/api');
vi.mock('@/modules/processes/api');
vi.mock('@/util');

interface Params {
  error: boolean;
}

interface ProcessesParams extends Params {
  processes: Processes[];
}

const noticeMock = {
  id: 1,
  text: 'Notice 1',
};

const processesMock = [
  {
    id: 'string',
    name: 'string',
    schoolYear: {
      year: 2021,
    },
    stages: [
      {
        endAt: '2020-10-19 23:59:00',
        id: '1',
        name: 'Test',
        startAt: '2020-10-14 00:00:00',
        status: 'CLOSED',
        type: 'REGISTRATION_RENEWAL',
      },
    ],
    grades: [
      {
        id: 27,
        name: 'NÃ­vel I',
      },
    ],
  },
];

const onMounted = async (notice: Notice, processesParams: ProcessesParams) => {
  vi.mocked(getCookie).mockReturnValueOnce('1');

  vi.mocked(NoticeApi.list).mockResolvedValueOnce(notice);

  if (processesParams.error) {
    vi.mocked(ProcessApi.list).mockRejectedValueOnce(processesParams.processes);
  } else {
    vi.mocked(ProcessApi.list).mockResolvedValueOnce(processesParams.processes);
  }

  const wrapper = mount(Index, {
    global: {
      plugins: [createTestingPinia()],
      directives: {
        tooltip: VTooltip,
      },
      stubs: ['router-link', 'protocol-consult', 'process-stages'],
      mocks: {
        $router: () => ({
          push: spyPush,
        }),
      },
    },
  });

  await wrapper.vm.$nextTick();

  await flushPromises();

  return wrapper;
};

describe('Index', () => {
  let wrapper: VueWrapper<InstanceType<typeof Index>>;

  beforeEach(async () => {
    wrapper = await onMounted(noticeMock, {
      error: false,
      processes: processesMock,
    });

    await wrapper.vm.$nextTick();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('the component exists', () => {
    expect(wrapper).toBeTruthy();
  });

  test('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('the initial variables must be loaded with the correct data', () => {
    expect(wrapper.vm.notice).toMatchObject(noticeMock);
  });

  test('if has notices, the notice component must be rendered', () => {
    expect(wrapper.vm.hasNotice).toBeTruthy();

    const noticeBlock = wrapper.find('[data-test="notice-block"]');

    expect(noticeBlock.exists()).toBeTruthy();
  });

  test("if has'nt notices, the notice component should not be rendered", async () => {
    wrapper = await onMounted({} as Notice, {
      error: false,
      processes: processesMock,
    });

    const noticeBlock = wrapper.find('[data-test="notice-block"]');

    expect(noticeBlock.exists()).toBeFalsy();
  });

  test('if has processes, the processes component must be rendered', () => {
    const processes = wrapper.findAll('[data-test="process-block"]');

    expect(processes.length).toBe(1);
  });

  test("if has'nt processes, the processes component should not be rendered", async () => {
    wrapper = await onMounted({} as Notice, {
      error: true,
      processes: undefined as unknown as Processes[],
    });

    const processes = wrapper.findAll('[data-test="process-block"]');

    expect(processes.length).toBe(0);
  });

  test('should show modal intro, if showVideoTutorial is true', async () => {
    vi.mocked(getCookie).mockReturnValueOnce(undefined);

    wrapper = await onMounted(noticeMock, {
      error: false,
      processes: processesMock,
    });

    expect(wrapper.vm.showVideoTutorial).toBeTruthy();
  });

  test("shouldn't show modal intro, if showVideoTutorial is false", async () => {
    expect(wrapper.vm.showVideoTutorial).toBeFalsy();
  });
});
