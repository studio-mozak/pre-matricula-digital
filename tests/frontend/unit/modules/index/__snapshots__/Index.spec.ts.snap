// Vitest Snapshot v1

exports[`Index > should match snapshot 1`] = `
<div>
  <div
    class="row mt-3"
  >
    <div
      class="container-sm col-10 col-lg-6 mt-4 pt-lg-4"
    >
      <div
        class="d-flex align-items-center mb-5"
      >
        <div
          class="font-muli"
          style="color: #003473;"
        >
          <div
            style="font-size: 24px;"
          >
            Some slogan
          </div>
          <div
            style="font-size: 34px;"
          >
            Some City
          </div>
        </div>
        <div
          class="ml-auto col-5"
        >
          <img
            alt="Some City"
            class="img-fluid"
            src="https://logo.com"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="row mt-3"
  >
    <div
      class="col-12 col-md-10 offset-md-1"
    >
      <protocol-consult-stub
        title="Consulta de inscrição"
      />
    </div>
  </div>
  <!--v-if-->
  <div
    class="row mt-4"
    data-test="notice-block"
  >
    <div
      class="col-12 col-md-10 offset-md-1"
    >
      <div
        class="x-card x-card--bordered"
      >
        <div
          class="x-card__section x-card__section--vert pl-4 pr-4 pt-3 p-0"
        >
          <h5
            class="text-h5 text-primary"
          >
            Avisos
          </h5>
        </div>
        <div
          class="x-card__section x-card__section--vert pl-4 pr-4 pb-3 p-0"
        >
          <div
            class="row"
          >
            <div
              class="col-12 text-break overfflow-hidden"
            >
              Notice 1
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-11 col-lg-2 font-hind offset-lg-10 pr-0 pr-md-4 text-right"
            >
              <router-link-stub
                to="/avisos"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="row mt-3"
  >
    <div
      class="container-sm col-12 col-md-10 offset-md-1 mt-4"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h2
            class="title mb-3"
          >
            Inscrições
          </h2>
        </div>
      </div>
      
      <div
        data-test="process-block"
      >
        <div
          class="row"
        >
          <div
            class="col-12 mt-3 mt-sm-2 d-flex"
          >
            <h4
              class="m-0"
            >
              string (2021) 
            </h4>
            <span
              class="ml-auto p-2 bg-gray-300 text-gray-700 rounded"
            >
               Lista de espera não disponível 
            </span>
          </div>
        </div>
        <process-stages-stub
          class="mb-3"
          process="[object Object]"
        />
      </div>
      
    </div>
  </div>
  <!--v-if-->
</div>
`;
